### PR TITLE
PM Prize Category Group API

### LINKED ISSUE #37: [PM] Develop, Test, and Document PM Prize Category Group API
**Describe the solution you'd like**
An API should be created to handle basic PM Prize Category Group (example: grand prizes, sponsor prizes, etc.) operations. This API should include the following endpoints:

- GET pm/groups/ - @ejub-has 
- POST pm/groups/ - @ejub-has 
- PATCH pm/groups/[groupId] - @trode21 
- DELETE pm/groups/[groupId] - @trode21 

**Additional context**
_PM Goal:_

The goal of this module is to provide a central area to distinguish different prizes for each category (sponsor or hackathon general prize). The way that this will mainly be used is that the organizers, and/or sponsors will have the ability to delete, update, and view prize categories that will be available for the hackathon. The specific sponsors will only be able to access/edit their own prize categories, meanwhile the organizers will have a view and be able edit all prize categories. The hackers, volunteers, and mentors will be able to just view the prize categories and the sponsor (if applicable) for them.

_Key Features in this Item:_

- Organizers can create super categories for prizes, and assign prizes to them (example: grand prizes, sponsor prizes, etc.)
- Organizers can update the display order for prizes within a super category
- Hackers, mentors, and volunteers can view the prize categories’ names, descriptions, eligibility, and prize name

### PR DESCRIPTION
**Pull request by ejub-has**
Closes #37

### CODE REVIEW

**Comment by nicksinishtaj** [comment]
Can you do the following for 93-95, each PR accordingly.

Copy paste this comment I have right below here, and then update with your endpoints and the data you're using to verify in the API explorer. Just makes it easier when referencing back to these what test data to run.

### Tests via API Explorer (ORGANIZER user):
**POST mewm/events/**

```JSON
{
    "eventId": "event1",
    "eventName": "Ejub's Excel Contest",
    "eventDescription": "Ejub displays his magnificent skill of beautiful art in Microsoft Excel. They refer to Ejub as the Picasso of Spreadsheets ",
    "virtual": true
}
```

> Status: 201

**GET mewm/events/event1**

mewm/events/event1

> Status: 200
> Body: {"eventId":"event1","virtual":true,"eventDescription":"Ejub displays his magnificent skill of beautiful art in Microsoft Excel. They refer to Ejub as the Picasso of Spreadsheets ","eventName":"Ejub's Excel Contest"}

**PATCH mewm/events/event1**

```JSON
{
	"eventName": "ejub excel"
}
```

> Status: 200

**GET mewm/events/event1**

> Status: 200
> Body: {"eventId":"event1","virtual":true,"eventDescription":"Ejub displays his magnificent skill of beautiful art in Microsoft Excel. They refer to Ejub as the Picasso of Spreadsheets ","eventName":"ejub excel"}

**POST mewm/events/**

```JSON
{
    "eventId": "event2",
    "eventName": "Ejub's LIVE Excel Contest",
    "eventDescription": "Ejub displays his magnificent skill of beautiful art in Microsoft Excel, but in real-time on a white board. They refer to Ejub as the Picasso of Spreadsheets ",
    "virtual": false
}
```

> Status: 201

**GET mewm/events/**

> Status: 200
> Body: {"miniEvents":[{"eventId":"event2","virtual":false,"eventDescription":"Ejub displays his magnificent skill of beautiful art in Microsoft Excel, but in real-time on a white board. They refer to Ejub as the Picasso of Spreadsheets ","eventName":"Ejub's LIVE Excel Contest"},{"eventId":"event1","virtual":true,"eventDescription":"Ejub displays his magnificent skill of beautiful art in Microsoft Excel. They refer to Ejub as the Picasso of Spreadsheets ","eventName":"ejub excel"}]}



**DELETE mewm/events/event1**

> Status: 200

**GET mewm/events/**

> Status: 200
> Body: {"miniEvents":[{"eventId":"event2","virtual":false,"eventDescription":"Ejub displays his magnificent skill of beautiful art in Microsoft Excel, but in real-time on a white board. They refer to Ejub as the Picasso of Spreadsheets ","eventName":"Ejub's LIVE Excel Contest"}]}

File: @types/modules/pm/groups.d.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+declare interface PMGroupUpdateRequest {
+    prizeGroupName?: string;
+    prizeGroupDescription?: string;
+    prizeGroupOrder?: Array<string>;
```
Comment (andrewdimmer) [inline @types/modules/pm/groups.d.ts:4]: Minor nitpick, but generally unless there is a really complicated type, it is generally easier to use `[]` to denote an array.
```suggestion
    prizeGroupOrder?: string[];
```

File: @types/modules/pm/groups.d.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  }
+
+  declare interface PMGroupCreateRequest extends PMGroupUpdateRequest {
+    prizeGroupId: string;
+    prizeGroupName: string;
+    prizeGroupDescription: string;
+    prizeGroupOrder: Array<string>;
```
Comment (andrewdimmer) [inline @types/modules/pm/groups.d.ts:11]: Minor nitpick, but generally unless there is a really complicated type, it is generally easier to use `[]` to denote an array. Also, might want to have this optional and autofill it upon creation, as there won't be any prizes when it gets created.
```suggestion
    prizeGroupOrder?: string[];
```

File: @types/modules/pm/groups.d.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  }
+
+  declare interface PMGroup extends PMGroupCreateRequest {
+    prizeGroupId: string;
+    prizeGroupName: string;
+    prizeGroupDescription: string;
+    prizeGroupOrder: Array<string>;
```
Comment (andrewdimmer) [inline @types/modules/pm/groups.d.ts:18]: Minor nitpick, but generally unless there is a really complicated type, it is generally easier to use `[]` to denote an array.
```suggestion
    prizeGroupOrder: string[];
```

File: @types/modules/mewm/companies.d.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
+declare interface STPMCompanyUpdateRequest {
```
Comment (andrewdimmer) [inline @types/modules/mewm/companies.d.ts:1]: Is this document required? I think this is a stray artifact and can be deleted.

File: @types/modules/stpm/prizeCatagories.d.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+type Prize = {
```
Comment (andrewdimmer) [inline @types/modules/stpm/prizeCatagories.d.ts:1]: Is this document required for this PR? I think this is a stray artifact and can be deleted.

File: functions/src/modules/pm/categories/prizes/deletePrize.ts [hunk 6]
```diff
@@ -1,5 +1,36 @@
-import { ExpressFunction } from "../../../../@types";
+import type { ExpressFunction } from "../../../../@types";
```
Comment (andrewdimmer) [inline functions/src/modules/pm/categories/prizes/deletePrize.ts:1]: Is this document required in this PR? I think this is a stray artifact and can be deleted.

File: functions/src/modules/pm/categories/prizes/getPrize.ts [hunk 7]
```diff
@@ -1,5 +1,62 @@
-import { ExpressFunction } from "../../../../@types";
+import type { ExpressFunction } from "../../../../@types";
```
Comment (andrewdimmer) [inline functions/src/modules/pm/categories/prizes/getPrize.ts:1]: Is this document required in this PR? I think this is a stray artifact and can be deleted.

File: functions/src/modules/pm/groups/createGroup.ts [hunk 8]
```diff
@@ -1,5 +1,54 @@
...
+        prizeGroupName: { rules: ["string"], required: true },
+        prizeGroupDescription: { rules: ["string"], required: true },
+        prizeGroupOrder: {
+          rules: [
+          {
+              type: "array",
+              rules: ["string"],
+          },
+          ],
+      required: true},
```
Comment (andrewdimmer) [inline functions/src/modules/pm/groups/createGroup.ts:29]: I think we should be able to get away with having this be optional, and then just autofill to an empty array if it is not given.

File: functions/src/modules/pm/groups/createGroup.ts [hunk 9]
```diff
@@ -1,5 +1,54 @@
...
+
+  const execute: ExpressFunction = (req, res, next) => {
+    const errorHandler = expressErrorHandlerFactory(req, res, next);
+  
+    const body = res.locals.parsedBody as PMGroupCreateRequest;
+
+  
+    firebaseApp
+      .firestore()
+      .collection("PrizeCategoryGroup")
```
Comment (andrewdimmer) [inline functions/src/modules/pm/groups/createGroup.ts:44]: Can we call this `prizeGroups` to stick with the naming convention from other modules?
```suggestion
      .collection("prizeGroups")
```

File: functions/src/modules/pm/groups/deleteGroup.ts [hunk 10]
```diff
@@ -1,5 +1,28 @@
...
+      })(req, res, next);
+};
 
-export default replaceMe;
+const execute: ExpressFunction = (req, res, next) => {
+    const errorHandler = expressErrorHandlerFactory(req, res, next);
+  
+    firebaseApp
+    .firestore()
+    .collection("PrizeCategoryGroup")
```
Comment (andrewdimmer) [inline functions/src/modules/pm/groups/deleteGroup.ts:18]: Can we call this `prizeGroups` to stick with the naming convention from other modules?
```suggestion
    .collection("prizeGroups")
```

File: functions/src/modules/pm/groups/getGroup.ts [hunk 11]
```diff
@@ -1,5 +1,25 @@
...
+import type { ExpressFunction } from "../../../@types";
+import { firebaseApp } from "../../../config/firebaseConfig";
+import { expressErrorHandlerFactory } from "../../../helpers";
 
-const replaceMe: ExpressFunction = (req, res, next) => {};
+const getGroup: ExpressFunction = (req, res, next) => { 
+   const errorHandler = expressErrorHandlerFactory(req, res, next);
+    firebaseApp
+    .firestore()
+    .collection("PrizeCategoryGroup")
```
Comment (andrewdimmer) [inline functions/src/modules/pm/groups/getGroup.ts:9]: Can we call this `prizeGroups` to stick with the naming convention from other modules?
```suggestion
    .collection("prizeGroups")
```

File: functions/src/modules/pm/groups/listGroups.ts [hunk 12]
```diff
@@ -1,5 +1,33 @@
...
+  })(req, res, next);
+};
 
-export default replaceMe;
+const execute: ExpressFunction = (req, res, next) => {
+  const errorHandler = expressErrorHandlerFactory(req, res, next);
+
+  firebaseApp
+    .firestore()
+    .collection("PrizeCategoryGroup")
```
Comment (andrewdimmer) [inline functions/src/modules/pm/groups/listGroups.ts:17]: Can we call this `prizeGroups` to stick with the naming convention from other modules?
```suggestion
    .collection("prizeGroups")
```

File: functions/src/modules/pm/groups/updateGroup.ts [hunk 13]
```diff
@@ -1,5 +1,49 @@
...
+    requestBodyTypeValidator(req, res, next)(validationRules, execute);
+}; 
 
-export default replaceMe;
+const execute: ExpressFunction = (req, res, next) => {
+  const errorHandler = expressErrorHandlerFactory(req, res, next);
+
+  firebaseApp
+    .firestore()
+    .collection("PrizeCategoryGroup")
```
Comment (andrewdimmer) [inline functions/src/modules/pm/groups/updateGroup.ts:39]: Can we call this `prizeGroups` to stick with the naming convention from other modules?
```suggestion
    .collection("prizeGroups")
```

File: functions/src/modules/pm/groups/createGroup.ts [hunk 14]
```diff
@@ -1,5 +1,54 @@
...
+         }, 
+    };
+    requestBodyTypeValidator(req, res, next)(validationRules, execute);
+  };
+
+  const execute: ExpressFunction = (req, res, next) => {
+    const errorHandler = expressErrorHandlerFactory(req, res, next);
+  
+    const body = res.locals.parsedBody as PMGroupCreateRequest;
+
```
Comment (andrewdimmer) [inline functions/src/modules/pm/groups/createGroup.ts:40]: Can we add a check to basically say if `prizeGroupOrder` is `undefined`, set this equal to an empty array?

**Review by andrewdimmer — CHANGES_REQUESTED** [review]
Awesome start! Few minor changes, then you should be good to go!

File: @types/modules/pm/groups.d.ts [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+    prizeGroupName?: string;
+    prizeGroupDescription?: string;
+    prizeGroupOrder?: string[];
+  }
+
+  declare interface PMGroupCreateRequest extends PMGroupUpdateRequest {
+    prizeGroupId: string;
+    prizeGroupName: string;
+    prizeGroupDescription: string;
+    prizeGroupOrder?: string[];
```
Comment (andrewdimmer) [inline @types/modules/pm/groups.d.ts:11]: Actually, this is redundant now due to inheritance...
```suggestion
```

File: functions/src/modules/pm/groups/createGroup.ts (hunk 8 above)
Comment (andrewdimmer) [inline functions/src/modules/pm/groups/createGroup.ts:29]: I think we should be able to get rid this.
```suggestion
      },
```

**Review by andrewdimmer — CHANGES_REQUESTED** [review]
Just a few more changes...

File: @types/modules/mewm/companies.d.ts [hunk 16]
```diff
@@ -4,7 +4,7 @@ declare interface STPMCompanyUpdateRequest {
   companyWebsite?: string;
   companyAcronym?: string;
   sponsorTierId?: string;
-  overriddenBenefits?: STPMTierUpdateRequest;
+  overriddenBenefits?: { [key: string]: string };
```
Comment (andrewdimmer) [inline @types/modules/mewm/companies.d.ts:7]:
```suggestion
  overriddenBenefits?: STPMTierUpdateRequest;
```

File: @types/modules/mewm/companies.d.ts [hunk 17]
```diff
@@ -14,7 +14,7 @@ declare interface STPMCompanyCreateRequest extends STPMCompanyUpdateRequest {
 }
 
 declare interface STPMCompany extends STPMCompanyCreateRequest {
-  overriddenBenefits: STPMTierUpdateRequest;
+ overriddenBenefits: { [key: string]: string };
```
Comment (andrewdimmer) [inline @types/modules/mewm/companies.d.ts:17]:
```suggestion
  overriddenBenefits: STPMTierUpdateRequest;
```

**Review by andrewdimmer — APPROVED** [review]
Provided the build tests pass, you should be all set to merge!

**Pull request merged**